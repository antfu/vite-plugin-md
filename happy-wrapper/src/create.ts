import { identity } from 'fp-ts/lib/function'
import { Text, Window } from 'happy-dom'
import { dasherize } from 'native-dash'
import { HappyMishap } from './errors'
import type { Container, HTML } from './happy-types'
import { isElement, isElementLike, isTextNodeLike } from './type-guards'
import { clone, solveForNodeType, toHtml } from './utils'
import type { Document, Fragment, IElement, IText } from './index'

/**
 * Converts an HTML string into a Happy DOM document tree
 */
export function createDocument(body: string, head?: string): Document {
  const window = new Window()
  const document = window.document
  document.body.innerHTML = body
  if (head)
    document.head.innerHTML = head
  return document
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type FragmentFrom<_T extends Container | 'html'> = Fragment

export function createFragment<C extends Container | HTML>(content?: C): FragmentFrom<C extends string ? 'html' : Fragment> {
  const window = new Window()
  const document = window.document
  const fragment = document.createDocumentFragment() as Fragment
  if (content)
    fragment.append(clone(content))

  return fragment as FragmentFrom<C extends string ? 'html' : Fragment>
}

/**
 * Creates a DOM Node which will be either an `IElement` or `IText` node
 * based on the content passed in.
 */
export const createNode = (node: Container | string): IElement | IText => {
  const frag = createFragment(node)
  if (isElementLike(frag))
    return frag.firstElementChild as IElement
  else if (isTextNodeLike(frag))
    return frag.firstChild as IText
  else
    throw new HappyMishap('call to createNode() couldn\'t be converted to IElement or IText node', { name: 'createNode()', inspect: node })
}

/**
 * Creates a IText Node
 */
export function createTextNode(text?: string): IText {
  if (!text) {
    console.warn('An empty string was passed into createTextNode(); will be ignored but probably a mistake')
    return new Text('')
  }

  const frag = createFragment(text)
  if (isTextNodeLike(frag))
    return frag.firstChild as unknown as IText
  else
    throw new HappyMishap(`The HTML passed in cannot be converted to a single text node: "${text}".`, { name: 'createFragment(text)', inspect: frag })
}

/**
 * Creates an element node and can preserve parent relationship if known
 */
export const createElement = (el: Container | HTML, parent?: IElement): IElement => solveForNodeType()
  .outputType<IElement>()
  .solver({
    node: (n) => {
      if (isElement(n))
        return createElement(n) as IElement
      else
        throw new HappyMishap('can\'t create an IElement from an INode node because it doesn\'t have a tagName property', { inspect: n })
    },
    html: (h) => {
      const frag = createFragment(h)
      if (isElementLike(frag)) {
        if (parent) {
          parent.append(frag.firstElementChild)
          return parent?.lastElementChild
        }

        return frag.firstElementChild
      }
      else { throw new HappyMishap('The HTML passed into createElement() is not convertible to a IElement node!', { name: 'createElement(html)', inspect: frag }) }
    },
    element: identity,
    text: (t) => {
      throw new HappyMishap('An IElement can not be created from a IText node because element\'s require a wrapping tag name!', { name: 'createElement(text)', inspect: t })
    },
    fragment: (f) => {
      if (isElement(f.firstElementChild))
        return f.firstElementChild as IElement

      else
        throw new HappyMishap(`Unable to create a IElement node from: \n\n${toHtml(f)}`, { name: 'createElement()' })
    },
    document: (d) => {
      if (isElementLike(d)) {
        if (parent)
          throw new HappyMishap('A Document and a parent IElement were passed into createElement. This is not a valid combination!')

        return d.firstElementChild
      }

      else { throw new HappyMishap('Can not create an Element from passed in Document', { name: 'createElement(document)', inspect: d }) }
    },
  })(el)

export interface CssVariable {
  prop: string
  value: string | number | boolean
}

export type ClassDefn = Record<string, string | boolean | number>
export type MultiClassDefn = [selector: string, keyValues: ClassDefn][]

export interface ClassApi {
  /** add a child selector */
  addChild: (selector: string, defn: ClassDefn) => ClassApi
  /** add CSS prop/values */
  addProps: (defn: ClassDefn) => ClassApi
}

export interface InlineStyle {
  /** add a single CSS variable (at a time); the CSS scope will ':root' unless specified */
  addCssVariable: (prop: string, value: string | number | boolean, scope?: string) => InlineStyle
  addClassDefinition: (selection: string, cb: ((api: ClassApi) => void)) => InlineStyle
  addCssVariables: (dictionary: Record<string, string>, scope?: string) => InlineStyle
  convertToVueStyleBlock: (lang: 'css' | 'scss', scoped: boolean) => InlineStyle

  finish: () => IElement
}

const renderClasses = (klasses: MultiClassDefn) => {
  return klasses.map(
    ([selector, defn]) => `\n\n  ${selector} {\n${Object.keys(defn).map(
      p => `    ${dasherize(p)}: ${defn[p]};`,
    ).join('\n')}\n  }`).join('\n')
}

/**
 * Creates a new `<style>...</style>` node and provides a simple API surface to allow
 * populating the contents with inline CSS content
 */
export const createInlineStyle = <T extends string = 'text/css'>(type: T = 'text/css' as T) => {
  const cssVariables: Record<string, CssVariable[]> = {}
  const cssClasses: MultiClassDefn = []
  let isVueBlock = false
  let isScoped = true
  let vueLang: 'css' | 'scss' = 'css'

  const api: InlineStyle = {
    addCssVariable(prop: string, value: string | number | boolean, scope = ':root') {
      if (!(scope in cssVariables))
        cssVariables[scope] = []
      cssVariables[scope].push({ prop: prop.replace(/^--/, ''), value })

      return api
    },
    addClassDefinition(selector, cb) {
      const classApi: ClassApi = {
        addChild: (child, defn) => {
          const childSelector = `${selector} ${child}`
          cssClasses.push([childSelector, defn])
          return classApi
        },
        addProps: (defn) => {
          cssClasses.push([selector, defn])
          return classApi
        },
      }
      cb(classApi)
      return api
    },
    addCssVariables(dictionary: Record<string, string | number | boolean>, scope = ':root') {
      Object.keys(dictionary).forEach(p => api.addCssVariable(p, dictionary[p], scope))

      return api
    },
    convertToVueStyleBlock(lang, scoped = true) {
      vueLang = lang
      isVueBlock = true
      isScoped = scoped

      return api
    },

    finish() {
      const setVariable = (scope: string, defn: Record<string, any>) => `${scope} {\n${Object.keys(defn).map(prop => `    --${defn[prop].prop}: ${defn[prop].value}${String(defn.prop).endsWith(';') ? '' : ';'}`).join('\n')}\n  }`

      let text = ''
      // variables
      Object.keys(cssVariables).forEach(
        (v) => {
          text = `${text}${setVariable(v, cssVariables[v])}\n`
        },
      )
      // classes
      text = `${text}${renderClasses(cssClasses)}`

      return isVueBlock
        ? createElement(`<style lang="${vueLang}"${isScoped ? ' scoped' : ''}>\n${text}\n</style>`)
        : createElement(`<style type="${type}">\n${text}\n</style>`)
    },
  }

  return api
}
