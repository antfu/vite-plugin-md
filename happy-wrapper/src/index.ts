export { Document, DocumentFragment as Fragment, IElement, IText } from 'happy-dom'

export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
