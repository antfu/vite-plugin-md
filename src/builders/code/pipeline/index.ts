export * from './addLanguage'
export * from './addClipboard'
export * from './convertBlocksToDomNodes'
export * from './defaultBlocks'
export * from './extractMarkdownItTokens'
export * from './expandCodeBlockVariables'
export * from './inlineStyles'
export * from './highlightLines'
export * from './renderHtml'
export * from './updateLineNumbers'
export * from './useHighlighter'
export * from './userRules'
export * from './updatePreWrapper'
export * from './updateFrontmatterWithCodeBlock'
export * from './updateCodeBlockWrapper'
