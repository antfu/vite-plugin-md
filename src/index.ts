import type { Plugin } from 'vite'
import { createFilter } from '@rollup/pluginutils'
import { createSfcComponent } from './createSfcComponent'
import { resolveOptions } from './options'
import type { Options } from './types'
export { link, meta, code } from './builders'

function VitePluginMarkdown(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)
  const markdownToVue = createSfcComponent(options)

  const filter = createFilter(
    userOptions.include || /\.md$/,
    userOptions.exclude,
  )

  let config: Parameters<Exclude<Plugin['configResolved'], undefined>>[0]

  return {
    name: 'vite-plugin-md',
    enforce: 'pre',
    configResolved(c) {
      config = { ...c }
    },
    transform(raw, id) {
      if (!filter(id))
        return
      try {
        return markdownToVue(config)(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function() {
        return markdownToVue(config)(ctx.file, await defaultRead())
      }
    },
  }
}

export default VitePluginMarkdown
