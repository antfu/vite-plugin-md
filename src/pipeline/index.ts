export * from './applyMarkdownItOptions'
export * from './composeSfcBlocks'
export * from './createParser'
export * from './convertHandlerToTask'
export * from './escapeCodeTagInterpolation'
export * from './extractBlocks'
export * from './extractFrontmatter'
export * from './finalize'
export * from './frontmatterPreprocess'
export * from './loadMarkdownItPlugins'
export * from './parse'
export * from './transformsBefore'
export * from './transformsAfter'
export * from './wrapHtml'
