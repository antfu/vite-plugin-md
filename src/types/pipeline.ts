import type MarkdownIt from 'markdown-it'
import type { MaybeRef } from '@vueuse/core'
import type * as TE from 'fp-ts/TaskEither'
import type { UserConfig } from 'vite'
import type { Either } from 'fp-ts/lib/Either'
import type { Fragment, IElement } from 'happy-wrapper'
import type { EnumValues, Frontmatter, MetaProperty, ResolvedOptions } from './core'

export enum PipelineStage {
  /**
   * Initialized with incoming filename, config, options,
   * available events (core and provided by builders).
   */
  initialize = 'initialize',

  /**
   * All frontmatter has been extracted from default values and page values
   * but no mapping has been done yet.
   *
   * Note: this is the event hook which the included `meta` builder connects
   * to and it in turn _provides_ a `metaMapped` hook.
   */
  metaExtracted = 'metaExtracted',

  /**
   * The **MarkdownIt** parser is initialized, all builders
   * connecting at this point will receive a valid `md` parser
   * object so that they can participate in MD-to-HTML parsing.
   */
  parser = 'parser',
  /**
   * The **MarkdownIt** parser is initialized and all builders
   * have been able to apply their customizations to it.
   */
  parsed = 'parsed',

  /**
   * The HTML has been converted to a HappyDom tree to allow
   * interested parties to manipulate contents using DOM based
   * queries.
   */
  dom = 'dom',

  /**
   * SFC blocks (template, script, and an array of customBlocks) are ready for
   * builders to inspect/mutate/etc.
   */
  sfcBlocksExtracted = 'sfcBlocksExtracted',

  /**
   * All mutations of page are complete; builders can hook into this stage but
   * will _not_ be able to mutate at this stage.
   */
  closeout = 'closeout',
}

export type IPipelineStage = EnumValues<PipelineStage>

export interface RulesUse {
  ruleName: string
  usage: 'adds' | 'patches' | 'modifies'
  description?: string
}

export type PipelineInitializer = (i?: Pipeline<PipelineStage.initialize>) => Pipeline<PipelineStage.initialize>

export interface BuilderRegistration<O extends BuilderOptions, S extends IPipelineStage> {
  name: string
  description?: string
  /** The lifecycle event/hook which this builder will respond to */
  lifecycle: S
  /**
   * The builder's handler function which receives the _payload_ for the
   * event lifecycle hook configured and then is allowed to mutate these
   * properties and pass back a similarly structured object to continue
   * on in that pipeline stage.
   */
  handler: BuilderHandler<O, S>

  /**
   * The options _specific_ to the builder
   */
  options: O

  /**
   * This isn't strictly required, but it is nice to express which rules you have used
   * modified, or added from the MarkdownIt parser.
   *
   * Note: builders should try to avoid mutating core rules; if they need a modification
   * for their purposes consider _monkey patching_ the rule so that downstream rules
   * have a better understanding of current rule state.
   */
  parserRules?: RulesUse[]

  /**
   * If this plugin needs to modify the configuration in some way at initialization
   * it can add a function here to do that. In most cases, the builder can simply
   * wait for their event hook to be called (at which point they will get the configuration
   * passed to them).
   */
  initializer?: BuilderHandler<O, PipelineStage.initialize>
}

export type Parser<S extends IPipelineStage> = S extends 'parser' | 'parsed' | 'dom' | 'sfcBlocksExtracted' | 'closeout'
  ? {
      /** the **MarkdownIT** parser instance */
      parser: MarkdownIt

    }
  : {}

/**
 * The **Route** custom-block can be configured with these properties.
 *
 * Note: this is best done with the _meta_ builder
 */
export interface RouteConfig {
  name?: string
  path?: string
  meta?: Record<string, any>
}

export interface LinkProperty {
  rel?: string
  href?: string
  integrity?: string
  crossorigin?: string
  [key: string]: unknown
}

export interface StyleProperty {
  type?: string
  [key: string]: unknown
}

/**
 * The <base> element specifies the base URL and/or target for all
 * relative URLs in a page.
 */
export interface BaseProperty {
  href?: string
  target?: string
}

export interface ScriptProperty {
  /**
   * For classic scripts, if the async attribute is present, then the classic
   * script will be fetched in parallel to parsing and evaluated as soon as it is
   * available.
   *
   * For module scripts, if the async attribute is present then the scripts and
   * all their dependencies will be executed in the defer queue, therefore they will
   * get fetched in parallel to parsing and evaluated as soon as they are available.
   */
  async?: boolean
  crossorigin?: string
  /** only to be used alongside the `src` attribute */
  defer?: boolean
  /** Provides a hint of the relative priority to use when fetching an external script.  */
  fetchPriority?: 'high' | 'low' | 'auto'

  integrity?: string
  /**
   * This Boolean attribute is set to indicate that the script should not be executed
   * in browsers that support ES2015 modules â€” in effect, this can be used to serve
   * fallback scripts to older browsers that do not support modular JavaScript code.
   */
  nomodule?: boolean
  nonce?: string
  referencePolicy?: 'no-referrer' | 'no-referrer-when-downgrade' | 'origin' | 'origin-when-cross-origin' | 'same-origin' | 'strict-origin' | 'strict-origin-when-cross-origin' | 'unsafe-url'
  src?: string
  type?: string
  [key: string]: unknown
}

/**
 * The properties destined for the HEAD block in the HTML
 */
export interface HeadProps {
  title?: MaybeRef<string>
  meta?: MaybeRef<MetaProperty[]>
  link?: MaybeRef<LinkProperty[]>
  base?: MaybeRef<BaseProperty[]>
  style?: MaybeRef<StyleProperty[]>
  script?: MaybeRef<ScriptProperty[]>
  htmlAttrs?: MaybeRef<Record<string, unknown>[]>
  bodyAttrs?: MaybeRef<Record<string, unknown>[]>
  [key: string]: unknown
}

export interface PipelineUtilityFunctions {
  /**
   * Adds a `<link>` to the page's header section
   */
  addLink: (link: LinkProperty) => void
  /**
   * Adds a `<script>` reference to the page's header section
   */
  addScriptReference: (script: ScriptProperty) => void

  /**
   * Allows the addition of code which will be brought into the
   * `<script setup>` block if using VueJS 3.x and into a normal
   * `<script>` block in VueJS2
   */
  addCodeBlock: (name: string, script: string, forVue2?: string[] | undefined) => void

  /**
   * Adds a `<style>` reference to the page's header section
   */
  addStyleReference: (style: StyleProperty) => void
  /**
   * Adds a VueJS `<script>` block to the HTML (which VueJS will eventually place in HEAD). A style block should be named so that downstream consumers
   * can -- potentially -- override or further modify the style.
   */
  addStyleBlock: (name: string, style: IElement | string) => void
}

export type MetaExtracted<S extends IPipelineStage> = S extends 'initialize'
  ? {}
  : {
      /** the frontmatter metadata */
      frontmatter: Frontmatter

      /**
       * The markdown content (after extracting frontmatter)
       */
      md: string

      /**
       * Meta properties that will be put into the HEAD section
       *
       */
      // TODO: this should be removed as it's duplicative with what's in `head`
      meta: MetaProperty[]

      excerpt?: string

    } & PipelineUtilityFunctions

export type HtmlContent<S extends IPipelineStage> = S extends 'parsed' | 'sfcBlocksExtracted' | 'closeout'
  ? {
      /**
       * the HTML produced from MD content (and using parser rules passed in)
       */
      html: string
    }
  : S extends 'dom'
    ? {
        /**
         * the HTML wrapped into a HappyDom fragment
         */
        html: Fragment

        /**
         * If any code blocks were found on the page then their languages will be represented
         * here as a `Set<string>`
         */
        fencedLanguages: Set<string>
      }
    : {}

export type Blocks<S extends IPipelineStage> = S extends 'sfcBlocksExtracted' | 'closeout'
  ? {
      /**
       * all hoisted scripts
       */
      hoistedScripts: string[]

      /** the SFC's template block (aka, html content) */
      templateBlock: string

      /** the `<script [setup] ...>` block */
      scriptBlock: string

      /** any other top-level SFC blocks besides "template" and "script" */
      customBlocks: string[]
    }
  : {}

export type Completed<S extends IPipelineStage> = S extends 'closeout'
  ? {
      /** the finalized component in string form */
      component: string
    }
  : {}

export type Pipeline<S extends IPipelineStage> = {
  fileName: string
  /** the raw content in the file being processed */
  content: string
  /** the `vite-plugin-md` options */
  options: ResolvedOptions
  /** the Vite config */
  viteConfig: UserConfig

  /**
   * All properties which are destined for the HEAD section of the HTML
   */
  head: HeadProps
  /**
   * Meta properties associated with a page's route; typically used
   * and managed with the "meta" builder.
   */
  routeMeta?: RouteConfig

  /**
   * Indicates which _languages_ were found on the page; this property typically
   * shouldn't be set but rather is managed by the `code()` builder and provided
   * for contextual decisions.
   */
  codeBlockLanguages: {
    /** the language setting the page author used */
    langsRequested: string[]
    /** the language used to parse with the highlighter */
    langsUsed: string[]
  }

  /**
   * A store for _named_ VueJS `<style>` blocks which will be injected
   * into the SFC component at the appropriate time.
   */
  vueStyleBlocks: Record<string, IElement>

  /**
   * Provides a _named_ set of code blocks which will be injected into
   * the VueJS's SFC's `<script setup>` section for Vue 3 and in a
   * `<script>` block. All blocks will be setup for Typescript.
   *
   * Note: contributors may optionally include additional trailing lines
   * for Vue2; this will allows you to export variables you've defined.
   */
  vueCodeBlocks: Record<string, string | [base: string, vue2Exports: string[]]>
} & Parser<S> & MetaExtracted<S> & HtmlContent<S> & Blocks<S> & Completed<S>

/**
 * The Builder's event listener/handler
 */
export type BuilderHandler<
  O extends BuilderOptions,
  S extends IPipelineStage,
> = (payload: Pipeline<S>, options: O) => Promise<Pipeline<S>>

/**
 * Users configure a `BuilderHandler` and we wrap this up functionality
 * with a higher-order `BuilderTask`.
 *
 * @returns TaskEither<string, Pipeline<S>>
 */
export type BuilderTask<
  S extends IPipelineStage,
> = () => (payload: Pipeline<S>) => TE.TaskEither<string, Pipeline<S>>

/**
 * Builder's must provide an export which meets this API constraint. Basic
 * structure of this higher order function is:
 *
 * - options( ) -> register( ) -> { handler( payload ) -> payload }
 */
export type BuilderApi<O extends {}, S extends IPipelineStage> = (options?: O) => () => BuilderRegistration<O, S>

export type InlineBuilder = <N extends string, L extends IPipelineStage>(name: N, lifecycle: L) => (payload: Pipeline<L>) => Pipeline<L>

/**
 * Builder options are expected to be a key/value dictionary but must
 * be allowed to be an empty object
 * */
export type BuilderOptions = Record<string, any> | {}

export type BuilderConfig = Record<IPipelineStage, BuilderRegistration<any, any>[] | []>

/**
 * Carries an Either<T> condition which is either:
 * - a _string_ error condition
 * - a `Pipeline<S>` value
 */
export type PipeEither<S extends IPipelineStage> = Either<string, Pipeline<S>>

/**
 * Carries an `TaskEither<T>` condition which is either:
 * - a _string_ error condition
 * - a `() => Promise<Pipeline<S>>` task
 */
export type PipeTask<S extends IPipelineStage> = TE.TaskEither<string, Pipeline<S>>

/**
 * A pipeline payload or either an async `PipeTask<S>` or a synchronous `PipeEither<S>`
 */
export type PipelinePayload<S extends IPipelineStage> = PipeTask<S> | PipeEither<S>

/**
 * A _synchronous_ transformer function which:
 *
 * - receives a payload of `PipeEither<F>`, and
 * - converts it to a type of `PipeEither<T>`
 */
export type SyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>

/**
* An _asynchronous_ transformer function which:
*
* - receives a payload of `PipeTask<F>` (async) or `PipeEither<F>` (sync)
* - converts it to a type of `PipeTask<T>`
*/
export type AsyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>
