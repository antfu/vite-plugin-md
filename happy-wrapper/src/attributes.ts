import { string } from 'fp-ts'
import { pipe } from 'fp-ts/lib/function'
import type { IElement, INode } from 'happy-dom'
import { createFragment } from './create'
import { HappyMishap } from './errors'
import type { Container, DocRoot, GetAttribute, HTML } from './happy-types'
import { isDocument, isElement, isFragment } from './type-guards'
import { getNodeType, solveForNodeType, toHtml } from './utils'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttribute<T extends string> = (value: string) => <N extends Container | HTML>(node: N) => N

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttributeTo<_T extends string, _V extends string> = <N extends Container | HTML>(node: N) => N

export const setAttribute = <T extends string>(
  attr: T,
): SetAttribute<T> => <V extends string>(
    value: V,
  ): SetAttributeTo<T, V> => <N extends Container | HTML>(
      node: N,
    ): N => {
    const invalidNode = (n: INode) => {
      throw new HappyMishap(`You can not use the setAttribute() utility on a node of type: "${getNodeType(n)}"`, { name: `setAttribute(${attr})(${value})(INVALID)` })
    }
    const result = solveForNodeType()
      .mirror()
      .solver({
        html: h => pipe(h, createFragment, f => setAttribute(attr)(value)(f), toHtml),
        text: t => invalidNode(t),
        node: n => invalidNode(n),
        fragment: (f) => {
          f.firstElementChild.setAttribute(attr, value)
          return f
        },
        document: (d) => {
          d.body.firstElementChild.setAttribute(attr, value)
          return d
        },
        element: (e) => {
          e.setAttribute(attr, value)
          return e
        },
      })(node)

    return result
  }

export const getAttribute = <T extends string>(attr: T): GetAttribute<T> => {
  return solveForNodeType('text', 'node')
    .outputType<string>()
    .solver({
      html: h => pipe(h, createFragment, getAttribute(attr)),
      fragment: f => f.firstElementChild.getAttribute(attr),
      document: doc => doc.body.firstElementChild.getAttribute(attr),
      element: el => el.getAttribute(attr),
    })
}

const getClass = getAttribute('class')
const setClass = setAttribute('class')
/**
 * Provides the classes defined on a given container's top level
 * element as an array of strings
 */
export const getClassList = (container: Container | HTML | null): string[] => {
  if (!container)
    return []

  return solveForNodeType().outputType<string[]>().solver({
    html: h => pipe(h, createFragment, getClassList),
    document: d => getClass(d.body.firstElementChild)?.split(/\s+/) || [],
    fragment: f => getClass(f.firstElementChild)?.split(/\s+/) || [],
    element: e => getClass(e)?.split(/\s+/) || [],
    text: (n) => { throw new HappyMishap('Passed in a text node to getClassList!', { name: 'getClassList', inspect: n }) },
    node: (n) => { throw new HappyMishap('Passed in an unknown node type to getClassList!', { name: 'getClassList', inspect: n }) },
  })(container).filter(i => i)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type RemoveClass<R extends string | string[]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Removes a class from the top level node of a container's body.
 *
 * Note: if the class wasn't present then no change is performed
 */
export const removeClass = <R extends string | string[]>(
  remove: R,
): RemoveClass<R> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const current = getClass(doc)?.split(/\s+/g) || []
  const toRemove: string[] = !Array.isArray(remove) ? [remove] : remove

  const resultantClassString = Array.from(
    new Set<string>(current.filter(c => !toRemove.includes(c))),
  )
    .filter(i => i)
    .join(' ')

  return setClass(resultantClassString)(doc)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type AddClass<A extends string[] | string[][]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Adds a class to the top level node of a document's body.
 */
export const addClass = <A extends string[] | string[][]>(
  ...add: A
): AddClass<A> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const toAdd = (Array.isArray(add) ? add.flat() : [add]) as string[]

  const currentClasses = getClass(doc)?.split(/\s+/g) || []
  const resultantClasses = Array.from(new Set<string>([...currentClasses, ...toAdd]))

  return setClass(resultantClasses.join(' ').trim())(doc) as D
}

export type Filter = (string | RegExp)
export type FilterCallback = (removed: string[]) => void
export type FiltersWithCallback = [FilterCallback, ...Filter[]]

function hasFilterCallback(filters: Filter[] | FiltersWithCallback): filters is FiltersWithCallback {
  return typeof filters[0] === 'function'
}

/**
 * Filters classes out from a given element using _filters_, where a filter:
 *
 * - string - when a string it will compare for a direct match
 * - RegExp - will run the RegExp's `test(class)` method
 *
 * Optionally you may pass in a callback function as the first parameter in the
 * the list and this callback will be then called with all filtered properties
 * passed to it. This is useful for creating desirable side-effects like _moving_
 * the classes to some other DOM element (for instance).
 */
export const filterClasses = <A extends Filter[] | [FilterCallback, ...Filter[]]>(
  ...args: A
) => <D extends DocRoot | IElement | HTML>(
    doc: D,
  ): D => {
  const el = isDocument(doc) || isFragment(doc)
    ? doc.firstElementChild as IElement
    : isElement(doc)
      ? doc as IElement
      : null
  if (!el)
    throw new HappyMishap('An invalid container was passed into filterClasses()!', { name: 'filterClasses', inspect: doc })

  const filters = hasFilterCallback(args) ? args.slice(1) as Filter[] : args as Filter[]
  const cb = hasFilterCallback(args) ? args[0] : undefined
  const classes = getClassList(el)
  const removed: string[] = []

  classes.forEach((klass) => {
    const matched = !filters.every(f =>
      typeof f === 'string'
        ? f.trim() !== klass.trim()
        : !f.test(klass),
    )
    if (matched)
      removed.push(klass)
  })

  setClass(classes.filter(k => !removed.includes(k)).join(' '))(doc)

  if (cb)
    cb(removed)

  return doc
}
