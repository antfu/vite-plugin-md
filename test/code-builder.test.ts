import { readFile } from 'fs/promises'
import { describe, expect, it } from 'vitest'
import { composeSfcBlocks } from '../src/pipeline'
import { code } from '../src/index'

async function getFixture(file: string): Promise<string> {
  const content = await readFile(`test/fixtures/${file}`, 'utf-8')
  return content
}

describe('code() builder', () => {
  it.only('valid language choice is rendered', async() => {
    const { templateBlock } = composeSfcBlocks(
      'test/fixtures/ts-code-block.md',
      await getFixture('multi-code-block.md'),
      { builders: [code()] },
    )
    expect(templateBlock).toMatchSnapshot()
    expect(templateBlock.includes('language-ts')).toBeTruthy()
    expect(
      templateBlock.includes('language-bash'),
      'while "bash" is the default language, it should not be falling back to that with a valid language detected',
    ).toBeFalsy()
  })
  it('"unknown language" fallback is used when language stated but not matched', async() => {
    const { templateBlock } = composeSfcBlocks(
      'test/fixtures/ts-code-block.md',
      await getFixture('unknown-code-block.md'),
      { builders: [code()] },
    )

    const langLine = templateBlock.split('\n').find(i => i.includes('language-'))
    expect(templateBlock.includes('language-bash'), `default fallback is "bash" but language line was:\n${langLine}`).toBeTruthy()
    expect(templateBlock.includes('language-xxx')).toBeFalsy()
  })
  it('line numbers are displayed when set', async() => {
    const { templateBlock } = composeSfcBlocks(
      'test/fixtures/ts-code-block.md',
      await getFixture('ts-code-block.md'),
      { builders: [code({ lineNumbers: true })] },
    )
    expect(templateBlock).toMatchSnapshot()
  })
  it.todo('line highlighting of single line works')
  it.todo('line highlighting of line range works')
})
