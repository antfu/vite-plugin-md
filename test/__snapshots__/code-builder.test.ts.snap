// Vitest Snapshot v1

exports[`code() builder > line numbers are displayed when set 1`] = `
"<div class=\\"markdown-body\\"><h1>Code Builder Test</h1>
<p>this test is intended to validate testing of code blocks</p>
<pre class='language-ts line-numbers-mode'><code v-pre>
type Valid = 'foo' | 'bar' | 'baz'
const t: Valid = 'foo'
</code></pre><p>The test is now complete; everyone should return to their desks.</p>
</div>"
`;

exports[`code() builder > valid language choice is rendered 1`] = `
"<div class=\\"markdown-body\\"><h1>Code Builder Test</h1>
<p>This test is intended to validate testing of code blocks; the first test is a typescript block.</p>
<pre class='language-ts foobar'><code v-pre>
type Valid = 'foo' | 'bar' | 'baz'
const t: Valid = 'foo'
</code></pre><p>But now we have some Rust code too:</p>
<pre class='language-rust foobar' style='classless'><code v-pre>
let count: usize = 100;
</code></pre><p>The test is now complete; everyone should return to their desks.</p>
</div>"
`;
